src/synapse-plugins/converter-plugin/converter-plugin-backend.vala

```diff
@@ -0,0 +1,360 @@
+/*
+* Copyright (c) 2010 Michal Hruby <michal.mhr@gmail.com>
+*               2022 elementary LLC. (https://elementary.io)
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU General Public
+* License as published by the Free Software Foundation; either
+* version 2 of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Michal Hruby <michal.mhr@gmail.com>
+*/
+
+namespace Synapse {
+    struct UnitMatch {
+        Unit unit; // Unit that matches in UNITS
+        SIPrefix prefix; // Prefix taken into account
+        int dimension; // Dimension taken into account
+
+        public string description () {
+            string dim = "";
+            if (dimension == 2) {
+                dim = _("squared");
+            } else if (dimension == 3) {
+                dim = _("cubed");
+            }
+
+            /// TRANSLATORS First %s SI prefix, Second %s unit name, Third %s dimension (blank, squared or cubed);
+            return _("%s%s %s").printf (prefix.prefix, unit.description, dim);
+        }
+
+        public double factor () { // Taking into account size, prefix and dimension
+            double factor = 1.0;
+            double size = unit.get_factor ();
+            for (int i = 0; i < dimension; i++) {
+                factor *= prefix.factor;
+                factor *= size;
+            }
+
+            return factor;
+        }
+    }
+
+    public struct ResultData {
+        double factor;
+        string from_description;
+        string to_description;
+    }
+
+    public class ConverterPluginBackend : Object {
+        public static ConverterPluginBackend get_instance () {
+            if (instance == null) {
+                instance = new ConverterPluginBackend ();
+            }
+
+            return instance;
+        }
+
+        private Regex? convert_regex = null;
+        private static ConverterPluginBackend instance = null;
+
+        construct {
+            /* The regex describes a string which *resembles* a unit conversion request in the form
+             * <number> <unit> => <unit>.
+             * Some restrictions are placed on the form of <unit> (letters maybe followed by a 2 or 3).
+             */
+            try {
+                convert_regex = new Regex (
+                    """^\d*.?\d*[a-zA-Z\/ ]+(2|3)?=>[a-zA-Z\/ ]+(2|3)?$""",
```

Comment:
Why `d*.?\d` and not simply `\d*`

Reply:
My understanding was that`\d` matched a "digit" not a decimal point? So I included a match for a single decimal point.  I'll check if it can be omitted. Or I guess I could use `[0-9.]`?  My knowledge of regex is a bit rudimentary.

Reply:
I tried it with just `\d*` and decimal points were not recognised.

Reply:
Done

Reply:
> So I included a match for a single decimal point.

Ahh. But a `.` in this context means any character. You'd need to escape it (`\.`) to just match with `.`'s. 
Just the `\d` should not match `.`'s, see: https://regex101.com/r/sy5EHo/1
I would use `[\d.]`. Note that inside `[]`'s the dot only matches with dots and no escape is needed.

Example with `[\d.]`: https://regex101.com/r/jfp9SF/1

Reply:
That's what I originally thought re `\d` but now all the test units are passing - even the ones including a decimal point even though I've taken out the match for the decimal point so I am not sure what is going on :shrug:  Anyway I'll put the decimal point regex back in according to your suggestion.